core/src/main/java/nu/yona/server/subscriptions/service/UserUpdateService.java

```diff
@@ -0,0 +1,315 @@
+		}
+	}
+
+	private Optional<ConfirmationCode> createConfirmationCodeIfNumberUpdated(UserDto user, UserDto originalUser,
+			User updatedUserEntity)
+	{
+		if (isMobileNumberDifferent(user, originalUser))
+		{
+			ConfirmationCode confirmationCode = createConfirmationCode();
+			updatedUserEntity.setMobileNumberConfirmationCode(confirmationCode);
+			return Optional.of(confirmationCode);
+		}
+		return Optional.empty();
+	}
+
+	private boolean isMobileNumberDifferent(UserDto user, UserDto originalUser)
+	{
+		return !user.getMobileNumber().equals(originalUser.getMobileNumber());
+	}
+
+	@Transactional
+	public UserDto updateUserCreatedOnBuddyRequest(UUID id, String tempPassword, UserDto user)
+	{
+		User originalUserEntity = userRetrievalService.getUserEntityById(id);
+		// security check: should not be able to replace the password on an existing user
+		Require.that(originalUserEntity.isCreatedOnBuddyRequest(), () -> UserServiceException.userNotCreatedOnBuddyRequest(id));
+
+		EncryptedUserData retrievedEntitySet = retrieveUserEncryptedData(originalUserEntity, tempPassword);
+		User savedUserEntity = saveUserEncryptedDataWithNewPassword(retrievedEntitySet, user);
+		sendConfirmationCodeTextMessage(savedUserEntity.getMobileNumber(), savedUserEntity.getMobileNumberConfirmationCode(),
+				SmsTemplate.ADD_USER_NUMBER_CONFIRMATION);
+		UserDto userDto = userRetrievalService.createUserDtoWithPrivateData(savedUserEntity);
+		logger.info("Updated user (created on buddy request) with mobile number '{}' and ID '{}'", userDto.getMobileNumber(),
+				userDto.getId());
+		return userDto;
+	}
+
+	private EncryptedUserData retrieveUserEncryptedData(User originalUserEntity, String password)
+	{
+		// use a separate crypto session to read the data based on the temporary password
+		try (CryptoSession cryptoSession = CryptoSession.start(Optional.of(password),
+				() -> userRetrievalService.doPreparationsAndCheckCanAccessPrivateData(originalUserEntity.getId())))
+		{
+			return retrieveUserEncryptedDataInNewCryptoSession(originalUserEntity);
+		}
+	}
+
+	private EncryptedUserData retrieveUserEncryptedDataInNewCryptoSession(User originalUserEntity)
+	{
+		MessageSource namedMessageSource = messageService.getMessageSource(originalUserEntity.getNamedMessageSourceId());
+		MessageSource anonymousMessageSource = messageService.getMessageSource(originalUserEntity.getAnonymousMessageSourceId());
+		EncryptedUserData userEncryptedData = new EncryptedUserData(originalUserEntity, namedMessageSource,
+				anonymousMessageSource);
+		userEncryptedData.loadLazyEncryptedData();
+		return userEncryptedData;
+	}
+
+	@Transactional
+	public void addBuddy(UserDto user, BuddyDto buddy)
+	{
+		Require.that(user != null && user.getId() != null, InvalidDataException::emptyUserId);
+		Require.that(buddy != null && buddy.getId() != null, InvalidDataException::emptyBuddyId);
+
+		updateUser(user.getId(), userEntity -> {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "user" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://10.127.10.107:9000/coding_rules#rule_key=squid%3AS2259)